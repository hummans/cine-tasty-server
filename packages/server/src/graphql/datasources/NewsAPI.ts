import { RESTDataSource } from 'apollo-datasource-rest';

import env from '../../config/environment';
import { Article } from '../../lib/types';

const BASE_URL = 'http://newsapi.org/v2';
const ENDPOINT = 'everything';
const STATUS_OK = 'ok';
const QUERY = 'cinema';

interface APIOutput {
  publishedAt: string;
  urlToImage: string;
  source: {
    id?: string;
    name: string;
  };
  author: string;
  title: string;
  url: string;
}

export interface INewsAPI {
  validateAPIOutput: (output: APIOutput) => boolean;
  parseArticle: (apiOutput: APIOutput) => Article;
  getAllArticles: () => Promise<Article[]>;
  getFromParam: () => string;
}

class NewsAPI extends RESTDataSource implements INewsAPI {
  constructor() {
    super();
    this.baseURL = BASE_URL;
  }

  parseArticle({
    publishedAt,
    urlToImage,
    source,
    author,
    title,
    url,
  }: APIOutput): Article {
    const { name, id } = source;

    return {
      image: urlToImage,
      id: id || url,
      source: name,
      publishedAt,
      author,
      title,
      url,
    };
  }

  getFromParam(): string {
    const today = new Date();

    const month = (today.getMonth() + 1).toString().padStart(2, '0');
    const year = today.getFullYear();
    const day = today
      .getDate()
      .toString()
      .padStart(2, '0');

    return `${year}-${month}-${day}`;
  }

  validateAPIOutput(output: APIOutput): boolean {
    return Object.entries(output).every(([, value]) => !!value);
  }

  async getAllArticles(): Promise<Article[]> {
    const from = this.getFromParam();

    const { status, articles } = await this.get(ENDPOINT, {
      apiKey: env.NEWS_API_KEY,
      q: QUERY,
      from,
    });

    if (status !== STATUS_OK) {
      return [];
    }

    const result = articles
      .filter((article: APIOutput) => this.validateAPIOutput(article))
      .map((article: APIOutput) => this.parseArticle(article));

    return result;
  }
}

export default NewsAPI;
