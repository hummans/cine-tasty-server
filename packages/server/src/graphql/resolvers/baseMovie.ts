import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';
import {
  BaseMovieGenre_IdsArgs as BaseMovieGenreIdsArgs,
  QueryResolvers,
  MediaType,
} from '../../lib/types';
import { MediaItem } from '../../types';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  BaseMovie: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseMovieGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Movie.toLowerCase(), language),
  },
};

export default resolvers;
