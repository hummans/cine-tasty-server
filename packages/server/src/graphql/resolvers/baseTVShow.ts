import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';
import {
  BaseTvShowGenre_IdsArgs as BaseTvShowGenreIdsArgs,
  QueryResolvers,
  MediaType,
} from '../../lib/types';
import { MediaItem } from '../../types';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  BaseTVShow: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseTvShowGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Tv.toLowerCase(), language),
  },
};

export default resolvers;
