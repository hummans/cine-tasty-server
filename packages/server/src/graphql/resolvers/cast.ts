import { QueryResolvers, MediaType } from '../../lib/types';

type CastType = {
  media_type: 'tv' | 'movie';
};

const resolveCastType = (cast: CastType): string | null => {
  if (cast.media_type === MediaType.Movie.toLowerCase()) {
    return 'CastMovie';
  }

  if (cast.media_type === MediaType.Tv.toLowerCase()) {
    return 'CastTV';
  }

  return null;
};

const resolveTypes: QueryResolvers = {
  Cast: {
    __resolveType(cast: CastType): string | null {
      return resolveCastType(cast);
    },
  },
};

export default resolveTypes;
