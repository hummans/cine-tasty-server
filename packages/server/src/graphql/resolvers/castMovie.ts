import { MediaItem } from '../../types';
import {
  QueryResolvers,
  MediaType,
  CastMovieGenre_IdsArgs as CastMovieGenreIdsArgs,
} from '../../lib/types';
import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  CastMovie: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: CastMovieGenreIdsArgs,
    ): Promise<string[]> => {
      return mediaGenres.getMediaGenres(
        genre_ids,
        MediaType.Movie.toLowerCase(),
        language,
      );
    },
  },
};

export default resolvers;
