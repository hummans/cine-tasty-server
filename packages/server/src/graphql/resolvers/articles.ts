import { QueryResolvers } from '../../lib/types';
import { Datasource } from '../datasources';
import { Article } from '../../lib/types';

const resolvers: QueryResolvers = {
  Query: {
    articles: (
      _: {},
      __: {},
      { dataSources }: Datasource,
    ): Promise<Article[]> => dataSources.news.getAllArticles(),
  },
};

export default resolvers;
