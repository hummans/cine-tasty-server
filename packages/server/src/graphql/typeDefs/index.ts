import ISO6391Language from './types/ISO6391Language';
import root from './root';
import baseMovie from './types/baseMovie';
import basePerson from './types/basePerson';
import baseTVShow from './types/baseTVShow';
import cast from './types/cast';
import castMovie from './types/castMovie';
import castTV from './types/castTVShow';
import articleLanguage from './types/articleLanguage';
import mediaType from './types/mediaType';
import article from './types/article';
import articleQueryResult from './types/articleQueryResult';
import peopleQueryResult from './types/peopleQueryResult';
import person from './types/person';
import searchResults from './types/searchResults';
import searchTypes from './types/searchTypes';
import trendingMovies from './types/trendingMovies';
import queries from './queries';

export default [
  root,
  trendingMovies,
  article,
  peopleQueryResult,
  searchTypes,
  articleQueryResult,
  articleLanguage,
  baseMovie,
  basePerson,
  baseTVShow,
  searchResults,
  cast,
  person,
  mediaType,
  castMovie,
  castTV,
  person,
  ISO6391Language,
  queries,
];
