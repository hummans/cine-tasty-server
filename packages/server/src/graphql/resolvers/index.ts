import articles from './articles';
import baseMovie from './baseMovie';
import baseTVShow from './baseTVShow';
import cast from './cast';
import castMovie from './castMovie';
import castTV from './castTV';
import people from './people';
import person from './person';
import personKnownFor from './personKnownFor';
import search from './search';
import searchResultItem from './searchResultItem';

export default [
  articles,
  baseMovie,
  baseTVShow,
  cast,
  castTV,
  castMovie,
  people,
  person,
  personKnownFor,
  search,
  searchResultItem,
];
