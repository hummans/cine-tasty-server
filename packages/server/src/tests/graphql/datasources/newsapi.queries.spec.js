/* eslint-disable @typescript-eslint/explicit-function-return-type */

import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';

import { rawArticleWithId } from './fixtures/getAllArticlesStub';

import NewsAPI from '../../../graphql/datasources/NewsAPI';
import resolvers from '../../../graphql/resolvers';
import typeDefs from '../../../graphql/typeDefs';

const GET_ARTICLES = gql`
  query GetArticles {
    articles {
      publishedAt
      source
      author
      title
      image
      url
      id
    }
  }
`;

const makeTestServer = () => {
  const newsAPI = new NewsAPI();

  const server = new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      news: newsAPI,
    }),
  });

  return { server, newsAPI };
};

describe('[Queries.NewsAPI]', () => {
  it('fetches an array of articles from the News API', async () => {
    const { server, newsAPI } = makeTestServer();

    newsAPI.get = jest.fn(() => ({
      articles: [rawArticleWithId],
      status: 'ok',
    }));

    const { query } = createTestClient(server);

    const result = await query({
      query: GET_ARTICLES,
    });

    expect(result).toMatchSnapshot();
  });

  it("returns an empty array when the status isn't ok", async () => {
    const { server, newsAPI } = makeTestServer();

    newsAPI.get = jest.fn(() => ({
      articles: [rawArticleWithId],
      status: 'error',
    }));

    const { query } = createTestClient(server);

    const result = await query({
      query: GET_ARTICLES,
    });

    expect(result).toMatchSnapshot();
  });
});
