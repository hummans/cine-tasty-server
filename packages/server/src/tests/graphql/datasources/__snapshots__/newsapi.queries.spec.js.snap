// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Queries.NewsAPI] fetches an array of articles from the News API 1`] = `
Object {
  "data": Object {
    "articles": Array [
      Object {
        "author": "author",
        "id": "id",
        "image": "urlToImage",
        "publishedAt": "publishedAt",
        "source": "name",
        "title": "title",
        "url": "url",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`[Queries.NewsAPI] returns an empty array when the status isn't ok 1`] = `
Object {
  "data": Object {
    "articles": Array [],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
